### PR TITLE
refine Folly recipe

### PR DESCRIPTION
**Pull request by kexianda**
**folly**

refine folly recipe
    
    1. add sse42 option
    some libraries depends on folly, but with simd enabled. It will causes
    compiling error if folly has no such option.
    
    2. folly project exports cmake file with components, such as folly::follybenchmark.
    To keep consistent with folly's project, add components info

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/folly/all/conanfile.py [hunk 1]
```diff
@@ -144,6 +150,9 @@ def source(self):
     def _configure_cmake(self):
         cmake = CMake(self)
         if can_run(self):
+            if self.options.with_sse4_2:
```
Comment (uilianries) [inline recipes/folly/all/conanfile.py:153]: Since when is it available? Is it backward compatible?

File: recipes/folly/all/conanfile.py [hunk 2]
```diff
@@ -22,10 +22,12 @@ class FollyConan(ConanFile):
     options = {
         "shared": [True, False],
         "fPIC": [True, False],
+        "with_sse4_2" : [True, False],
     }
     default_options = {
         "shared": False,
         "fPIC": True,
+        "with_sse4_2" : True
```
Comment (uilianries) [inline recipes/folly/all/conanfile.py:30]:
```suggestion
        "use_sse4_2" : True
```

with is not a good prefix, as it's not a dependency. Prefer `use` instead: https://github.com/conan-io/conan-center-index/blob/master/docs/packaging_policy.md#recommended-feature-options-names

File: recipes/folly/all/conanfile.py [hunk 3]
```diff
@@ -254,3 +263,30 @@ def package_info(self):
...
+            self.cpp_info.components["follybenchmark"].set_property("pkg_config_name", "libfollybenchmark")
+            self.cpp_info.components["follybenchmark"].libs = ["follybenchmark"]
+            self.cpp_info.components["follybenchmark"].requires = ["libfolly"]
+
+            self.cpp_info.components["folly_test_util"].set_property("cmake_target_name", "Folly::folly_test_util")
+            self.cpp_info.components["folly_test_util"].set_property("pkg_config_name", "libfolly_test_util")
+            self.cpp_info.components["folly_test_util"].libs = ["folly_test_util"]
+            self.cpp_info.components["folly_test_util"].requires = ["libfolly"]
+
+            if Version(self.version) >= "2020.08.10.00" and self.settings.os == "Linux":
```
Comment (uilianries) [inline recipes/folly/all/conanfile.py:278]: Bad indentation?

Comment (kexianda) [reply recipes/folly/all/conanfile.py:278]: > 

not bad indentation.

File: recipes/folly/all/conanfile.py (hunk 2 above)
Comment (kexianda) [reply recipes/folly/all/conanfile.py]: thanks. fixed.

File: recipes/folly/all/conanfile.py (hunk 3 above)
Comment (kexianda) [reply recipes/folly/all/conanfile.py]: it is OK to adjust the indentation..

File: recipes/folly/all/conanfile.py (hunk 1 above)
Comment (kexianda) [reply recipes/folly/all/conanfile.py]: > 

sse4.2 support was in folly long long ago (>7 years)
I will set option default to False to be backward compatible.

File: recipes/folly/all/conanfile.py [hunk 4]
```diff
@@ -144,6 +151,9 @@ def source(self):
     def _configure_cmake(self):
         cmake = CMake(self)
         if can_run(self):
+            if self.options.use_sse4_2 and str(self.settings.arch) in ['x86', 'x86_64']:
```
Comment (uilianries) [inline recipes/folly/all/conanfile.py:154]:
```suggestion
            if self.options.get_safe("use_sse4_2"):
```

Let's say we are using ARM arch here. The option use_sse4_2 will not be available and will break conan create command. Checking the arch should be done in `validate()` method, like:

```python
def validate(self):
    if self.options.get_safe("use_sse4_2") and str(self.settings.arch) not in ['x86', 'x86_64']:
        raise ConanInvalidConfiguration(f"{self.ref} can use the option use_sse4_2 only on x86 and x86_64 archs.")

Comment (kexianda) [reply recipes/folly/all/conanfile.py]: > 

Thanks.   fixed.

**Comment by kexianda** [comment]
@uilianries  
"continuous-integration/jenkins/pr-merge — This commit cannot be built" [Details]
I've built it sucessfully on my linux desktop. I want to know how it failed, I clicked the 'Details', but got "403 Forbidden"
How to get detailed error information?  Thanks in advance.

**Comment by uilianries** [comment]
@kexianda The CI result says: https://github.com/conan-io/conan-center-index/pull/13621#issuecomment-1286502765

Please, remove the line /test_v1_package/conanfile.py:# pylint: skip-file

**Review by jwillikers — APPROVED** [review]

File: recipes/folly/all/conanfile.py [hunk 5]
```diff
@@ -103,7 +110,7 @@ def validate(self):
                 raise ConanInvalidConfiguration("{} requires C++{} support. The current compiler {} {} does not support it.".format(
                     self.name, self._minimum_cpp_standard, self.settings.compiler, self.settings.compiler.version))
 
-        if self.version < "2022.01.31.00" and self.settings.os != "Linux":
+        if Version(self.version) < "2022.01.31.00" and self.settings.os != "Linux":
             raise ConanInvalidConfiguration("Conan support for non-Linux platforms starts with Folly version 2022.01.31.00")
 
         if self.settings.os == "Macos" and self.settings.arch != "x86_64":
```
Comment (prince-chrismc) [inline recipes/folly/all/conanfile.py:116]: Can this finally be updated?

File: recipes/folly/all/conanfile.py [hunk 6]
```diff
@@ -254,3 +268,30 @@ def package_info(self):
         self.cpp_info.components["libfolly"].names["cmake_find_package_multi"] = "folly"
         self.cpp_info.components["libfolly"].set_property("cmake_target_name", "Folly::folly")
         self.cpp_info.components["libfolly"].set_property("pkg_config_name", "libfolly")
+
+        if Version(self.version) >= "2019.10.21.00":
+            self.cpp_info.components["follybenchmark"].set_property("cmake_target_name", "Folly::follybenchmark")
```
Comment (prince-chrismc) [inline recipes/folly/all/conanfile.py:273]: Where does this come from 🤔 looking at the latest it's not there https://github.com/facebook/folly/search?q=follybenchmark

🤢 

https://github.com/facebook/folly/blob/8874256376d2f8a32867f17c9472a446d6707604/folly/CMakeLists.txt#L22 


at some point this needs to be removed

**Review by prince-chrismc — COMMENTED** [review]
The new targets are not exported by the project (usually when it changes over versions we just do latest) but some more work is needed if you want this available in older versions

File: recipes/folly/all/conanfile.py (hunk 6 above)
Comment (kexianda) [reply recipes/folly/all/conanfile.py:273]: > Where does this come from 🤔 looking at the latest it's not there https://github.com/facebook/folly/search?q=follybenchmark
> 
> 🤢
> 
> https://github.com/facebook/folly/blob/8874256376d2f8a32867f17c9472a446d6707604/folly/CMakeLists.txt#L22
> 
> at some point this needs to be removed

hi @prince-chrismc  
follybenchmark target is still there : https://github.com/facebook/folly/blob/main/folly/CMakeLists.txt#L23

File: recipes/folly/all/conanfile.py (hunk 5 above)
Comment (kexianda) [reply recipes/folly/all/conanfile.py:116]: > Can this finally be updated?
thanks. fixed now

**Review by jwillikers — APPROVED** [review]

File: recipes/folly/all/conanfile.py (hunk 6 above)
Comment (prince-chrismc) [reply recipes/folly/all/conanfile.py:283]: I clearly did not drink enough coffee.

Perfect thank you for pointign that out!

**Review by prince-chrismc — APPROVED** [review]

**Pull request merged**